README.md
# perl-Data-CouchDB

[![Build Status](https://travis-ci.org/binary-com/perl-Data-CouchDB.svg?branch=master)](https://travis-ci.org/binary-com/perl-Data-CouchDB)
[![codecov](https://codecov.io/gh/binary-com/perl-Data-CouchDB/branch/master/graph/badge.svg)](https://codecov.io/gh/binary-com/perl-Data-CouchDB)


Changes
{{$NEXT}}
   switch to dzil
0.14


lib/Data/CouchDB.pm
package Data::CouchDB;

use 5.006;
use strict;
use warnings;

=head1 NAME

Data::CouchDB - CouchDB document management

=cut

our $VERSION = '0.14';

=head1 NAME

Data::CouchDB

=head1 SYNOPSYS

    my $couchdb = Data::CouchDB->new(
        replica_host => 'localhost',
        replica_port => 5432,
        master_host  => 'localhost',
        master_port  => 5432,
        couch        => 'testdb',
    );

=head1 DESCRIPTION

This class represents couchdb as a datasource.

=head1 ATTRIBUTES

=cut

use Moose;
use MooseX::StrictConstructor;
use namespace::autoclean;

use Cache::RedisDB;
use Data::CouchDB::Connection;
use Try::Tiny;
use LWP::UserAgent;

has name => (
    is  => 'ro',
    isa => 'Str'
);

=head2 db

db with which to operate.

=cut

has db => (
    is      => 'ro',
    isa     => 'Str',
    default => 'db',
);

=head2 couchdb

password for "couchdb" user

=cut

has couchdb => (
    is      => 'ro',
    default => 'TESTPASS',
);

=head2 replica_host

name of the host to read from.

=cut

has 'replica_host' => (
    is      => 'ro',
    default => 'localhost',
);

=head2 replica_port

port number in replica_host through which we can read.

=cut

has replica_port => (
    is      => 'ro',
    default => '5984',
);

=head2 replica_protocol

 protcol used to read.

=cut

has replica_protocol => (
    is  => 'ro',
    isa => 'Str',
);

=head2 master_host

name of the host to write to

=cut

has 'master_host' => (
    is      => 'ro',
    default => 'localhost',
);

=head2 master_port

port number in master_host through which we can write.

=cut

has master_port => (
    is      => 'ro',
    default => '5984',
);

=head2 master_protocol

protocol used to write.

=cut

has master_protocol => (
    is  => 'ro',
    isa => 'Str',
);

=head2 replica

The internal ds used to read from couchdb

=cut

has replica => (
    is         => 'ro',
    isa        => 'Data::CouchDB::Connection',
    lazy_build => 1,
);

=head2 master

The internal ds used to write to couchdb

=cut

has master => (
    is         => 'ro',
    isa        => 'Data::CouchDB::Connection',
    lazy_build => 1,
);

=head2 ua

Optionally passed ua(user_agent). If not passed the couchdb's default ua is used.

=cut

has ua => (
    is  => 'ro',
    isa => 'Maybe[LWP::UserAgent]',
);

=head1 METHODS

=head2 document

Get or set a couch document.

Usage,
    To get a document
        $couchdb->document($doc_id);

    To set a document
        $couchdb->document($doc_id, $data);

        $data is a HashRef


=cut

my $cache_namespace = 'COUCH_DOCS';

sub document {
    my $self = shift;
    my $doc  = shift;
    my $data = shift;

    my $cache_key = $self->db . '_' . $doc;
    if ($data) {
        Cache::RedisDB->del($cache_namespace, $cache_key) if ($self->_can_cache);
        $data = $self->master->document($doc, $data);
    } else {
        $data = Cache::RedisDB->get($cache_namespace, $cache_key) if ($self->_can_cache);

        if (not $data) {
            $data = $self->replica->document($doc);
            Cache::RedisDB->set($cache_namespace, $cache_key, $data, 127)
                if ($data and $self->_can_cache);
        }
    }

    return $data;
}

=head2 view

Query a couchdb view

Usage,
    Without Parameters
        $couchdb->view($db, $viewname);

    With Parameters
        $couchdb->view($db, $viewname, $parameters);

        $parameters is a HashRef


=cut

sub view {
    my $self   = shift;
    my $view   = shift;
    my $params = shift;

    return $self->replica->view($view, $params);
}

=head2 document_present

A syntatic sugar to check if a document

Usage,
    if ($couchdb->document_present($doc_id)) {
        ....
    }

Throws,
    Nothing

Returns,
    1     - if document is found.
    undef - if document is not found.

=cut

sub document_present {
    my $self = shift;
    my $doc  = shift;

    try { $self->replica->document($doc); } or return;

    return 1;
}

=head2 create_document

Creates a couch document

Usage,
    my $doc_id = $couchdb->create_document($doc_id);

=cut

sub create_document {
    my $self = shift;
    my $doc  = shift;

    return $self->master->create_document($doc);
}

=head2 delete_document

Deletes a couch document

Usage,
    $couchdb->delete_document($doc_id);

=cut

sub delete_document {
    my $self = shift;
    my $doc  = shift;

    return $self->master->delete_document($doc);
}

=head2 create_database

Creates a CouchDB Database.

Usage,
    $couchdb->create_database();

=cut

sub create_database {
    my $self = shift;
    return $self->master->create_database();
}

=head2 can_read

Confirms that you can read from this couchdb

Usage,
    if($couchdb->can_read) {
        ...
    }

Returns,
    1     - can read
    undef - otherwise
=cut

sub can_read {
    my $self = shift;
    return $self->replica->can_connect;
}

=head2 can_write

Confirms that you can write to this couchdb

Usage,
    if($couchdb->can_write) {
        ...
    }

Returns,
    1     - can write
    undef - otherwise

=cut

sub can_write {
    my $self = shift;
    return $self->master->can_connect;
}

sub _build_replica {
    my $self   = shift;
    my $params = {};

    $params->{host} = $self->replica_host;
    $params->{port} = $self->replica_port;
    $params->{db}   = $self->db;

    $params->{protocol} = $self->replica_protocol if ($self->replica_protocol);
    $params->{couchdb}  = $self->couchdb          if ($self->couchdb);

    return Data::CouchDB::Connection->new(%$params);
}

sub _build_master {
    my $self   = shift;
    my $params = {};

    $params->{host} = $self->master_host;
    $params->{port} = $self->master_port;
    $params->{db}   = $self->db;

    $params->{protocol} = $self->master_protocol if ($self->master_protocol);
    $params->{couchdb}  = $self->couchdb         if ($self->couchdb);

    return Data::CouchDB::Connection->new(%$params);
}

has '_can_cache' => (
    is         => 'ro',
    lazy_build => 1,
);

sub _build__can_cache {
    return try { Cache::RedisDB::redis_connection(); 1; };
}

__PACKAGE__->meta->make_immutable;


=head1 AUTHOR

Binary.com, C<< <support at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-data-couchdb at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-CouchDB>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Data::CouchDB


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-CouchDB>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Data-CouchDB>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Data-CouchDB>

=item * Search CPAN

L<http://search.cpan.org/dist/Data-CouchDB/>

=back


=head1 ACKNOWLEDGEMENTS

=cut

1; # End of Data::CouchDB


lib/Data/CouchDB/Handler.pm
package Data::CouchDB::Handler;

use Moose;

## VERSION

=head1 NAME

Data::CouchDB::Connection

=head1 METHODS

=cut

use Data::CouchDB;
use YAML::XS;
use feature 'state';

has couchdb_databases => (
    is         => 'rw',
    isa        => 'HashRef',
    #default    => sub { {} },
);

=head2 couchdb

=cut

sub couchdb {
    my $self    = shift;
    my $db      = shift;
    my $ua      = shift;
    my $db_name = $self->couchdb_databases->{$db} || $db;

    if ($ua or not defined $self->_couch_cache->{$db_name}) {
        my $params = $self->_build_couchdb_params($db_name);

        # For normal usage we dont expect to pass ua(user_agent).
        # So dont cache if ua is passed.
        if ($ua) {
            $params->{ua} = $ua;
            return Data::CouchDB->new(%$params);
        }

        $self->_couch_cache->{$db_name} = Data::CouchDB->new(%$params);
    }

    return $self->_couch_cache->{$db_name};
}

has couchdb => (
    is     => 'ro',
    reader => '_couchdb_configuration'
);

sub _build_couchdb_params {
    my $self    = shift;
    my $db_name = shift;

    my $params->{db} = $db_name;

    state $config = YAML::XS::LoadFile('/etc/couchdb.yml');
    if ($config->{master}->{ip} ne '127.0.0.1') {
        $params->{master_host}     = $config->{master}->{ip};
        $params->{master_port}     = 6984;
        $params->{master_protocol} = 'https://';
    }
    $params->{couchdb} = $config->{password};

    return $params;
}

has '_couch_cache' => (
    is      => 'ro',
    isa     => 'HashRef',
    default => sub { {}; },
);

__PACKAGE__->meta->make_immutable;

1;



lib/Data/CouchDB/Exceptions.pm
package Data::CouchDB::Exceptions;

use 5.006;
use strict;
use warnings;

## VERSION

=head1 NAME

Data::CouchDB - CouchDB document management

=cut

use Exception::Class (

    'Data::CouchDB::Exception' => {
        fields => ['uri', 'error'],
        isa    => 'Exception::Class::Base'
    },

    'Data::CouchDB::ConnectionFailed' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'Connection to couch db failed',
    },

    'Data::CouchDB::DBNotFound' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'The DB you are looking for does not exist',
        fields      => ['uri', 'error', 'db'],
    },

    'Data::CouchDB::RetrieveFailed' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'Document is not found in couch',
        fields      => ['db', 'uri', 'error', 'document'],
    },

    'Data::CouchDB::RevisionNotMatched' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'Revision of document provided not match the one in couch',
        fields      => ['db', 'uri', 'error', 'document'],
    },

    'Data::CouchDB::UpdateFailed' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'Revision of document provided not match the one in couch',
        fields      => ['db', 'uri', 'error', 'document'],
    },

    'Data::CouchDB::QueryFailed' => {
        isa         => 'Data::CouchDB::Exception',
        description => 'Revision of document provided not match the one in couch',
        fields      => ['db', 'uri', 'error', 'view'],
    },
);

=head1 METHODS

=head2 full_message

=cut

sub full_message { my $self = shift; return 'Connection to db ' . $self->db . '@' . $self->uri . ' failed with error ' . $self->error; }

1;


lib/Data/CouchDB/Role/VersionedData.pm
package Data::CouchDB::Role::VersionedData;
use strict; use warnings;

## VERSION

=head1 NAME

Data::CouchDB::Role::VersionedData

=head1 DESCRIPTION

Handles pulling in symbol-related data versioned by date.

=cut

use Net::SSL;
use JSON qw(to_json from_json);
use List::Util qw(min max);
use Moose::Role;
use Carp;
use URL::Encode qw(:all);
use Try::Tiny;

use Date::Utility;
use Data::CouchDB::Handler;

requires 'symbol', '_data_location', '_document_content';

=head1 ATTRIBUTES

=head2 for_date

The date for which we wish data

=cut

has for_date => (
    is      => 'ro',
    isa     => 'Maybe[Date::Utility]',
    default => undef,
);

=head2 current_document_id

=cut

has current_document_id => (
    is         => 'ro',
    isa        => 'Str',
    lazy_build => 1,
);

sub _build_current_document_id {
    my $self = shift;
    return $self->symbol;
}

=head2 document

The CouchDB document that this object is tied to.

=cut

has document => (
    is         => 'rw',
    lazy_build => 1,
);

sub _build_document {
    my $self = shift;

    my $document = $self->_couchdb->document($self->current_document_id);
    if ($self->for_date and $self->for_date->datetime_iso8601 lt $document->{date}) {
        my $params = {
            startkey   => [$self->symbol, $self->for_date->datetime_iso8601],
            endkey     => [$self->symbol],
            descending => 1,
            limit      => 1,
        };

        my $document_id = $self->_couchdb->view('by_date', $params)->[0];
        if ($document_id) {
            $document = $self->_couchdb->document($document_id);
        } else {
            $params = {
                startkey => [$self->symbol],
                endkey   => [$self->symbol, {}],
                limit    => 1,
            };

            $document_id = $self->_couchdb->view('by_date', $params)->[0];
            if ($document_id) {
                $document = $self->_couchdb->document($document_id);
            }

            $document //= {};
            $document->{date} = $self->for_date->datetime_iso8601;
        }
    }

    return $document;
}

has _couchdb => (
    is         => 'ro',
    isa        => 'Data::CouchDB',
    lazy_build => 1,
);

sub _build__couchdb {
    my $self = shift;

    my $handler = Data::CouchDB::Handler->new();
    if (exists $ENV{COUCHDB_DATABASES}) {
        $handler->couchdb_databases($ENV{COUCHDB_DATABASES});
    } else {
        $handler->couchdb_databases({
            $self->_data_location => $self->_data_location,
        });
    }

    return $handler->couchdb($self->_data_location);
}

=head2 add_to_history

Add the current live document to the searchable history

=cut

sub add_to_history {
    my ($self) = @_;

    # If its already a historical document hopefully its id will look like a uuid.
    if ($self->_looks_like_uuid($self->document->{_id})) {
        croak 'Saving historical document not permitted.';
    }

    delete $self->document->{_rev};
    my $historical_doc = $self->_couchdb->create_document();
    return $self->_couchdb->document($historical_doc, $self->document);
}

=head2 save

=cut

sub save {
    my $self = shift;

    if (not $self->_couchdb->document_present($self->current_document_id)) {
        $self->_couchdb->create_document($self->current_document_id);
    } else {
        $self->add_to_history;
    }

    my $new_document = $self->_document_content;
    $self->document($new_document);
    return $self->_couchdb->document($self->current_document_id, $new_document);
}

sub _looks_like_uuid {
    my ($self, $candidate) = @_;
    return ($candidate and $candidate =~ /^\w{32}$/);
}

1;


lib/Data/CouchDB/Connection.pm
package Data::CouchDB::Connection;

use 5.006;
use strict;
use warnings;

## VERSION

=head1 NAME

Data::CouchDB::Connection

=head1 SYNOPSYS

    my $couch_connection = Data::CouchDB::Connection->new(
        host   => 'localhost',
        port   => 5432,
        protocol => 'http://',
        couch => 'testdb',
    );

=head1 DESCRIPTION

This class represents a couchdb connection.

=head1 ATTRIBUTES

=cut

use Moose;
use MooseX::StrictConstructor;
use namespace::autoclean;

use CouchDB::Client;
use Net::SSL;
use LWP::UserAgent;
use HTTP::Request;
use HTTP::Headers;
use Try::Tiny;
use IO::Socket::SSL qw( SSL_VERIFY_NONE );
use Data::CouchDB::Exceptions;

=for Pod::Coverage SSL_VERIFY_NONE

=head2 host

name of the host to connect to.

=cut

has 'host' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

=head2 port

name of the port to connect to.

=cut

has 'port' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

=head2 db

db with which to operate.

=cut

has db => (
    is      => 'ro',
    isa     => 'Str',
    default => 'db',
);

=head2 protocol

protocol with which to connect.

Can be,
    'http://' or 'https://'

=cut

has 'protocol' => (
    is         => 'ro',
    isa        => 'Str',
    lazy_build => 1,
);

=head2 couchdb

password for "couchdb" user

=cut

has 'couchdb' => (
    is      => 'ro',
    isa     => 'Str',
    default => 'TESTPASS'
);

=head2 design_doc

Location inside the db where the design document is present.

=cut

has 'design_doc' => (
    is      => 'ro',
    default => '_design/docs'
);

=head2 uri

The uri with which we connect to.

=cut

has 'uri' => (
    is         => 'ro',
    isa        => 'Str',
    lazy_build => 1,
);

=head2 log_uri

The uri we use for throwing exceptions.

=cut

has 'log_uri' => (
    is         => 'ro',
    isa        => 'Str',
    lazy_build => 1,
);

=head2 ua

Optionally passed ua(user_agent). If not passed the couchdb's default ua is used.

=cut

has 'ua' => (
    is         => 'ro',
    lazy_build => 1,
);

=head1 METHODS

=head2 document

Get or set a couch document.

Usage,
    To get a document
        $couchdb->document($db, $doc_id);

    To set a document
        $couchdb->document($db, $doc_id, $data);

        $data is a HashRef

=cut

sub document {
    my $self      = shift;
    my $doc       = shift;
    my $data      = shift;

    return unless ($doc);
    my $couch_doc = $self->_doc($doc);
    try { $couch_doc->retrieve; }
    catch {
        Data::CouchDB::RetrieveFailed->throw(
            db       => $self->db,
            uri      => $self->log_uri,
            document => $doc,
            error    => $_,
        );
    };

    if ($data) {
        if (exists $data->{_rev}) {
            if ($data->{_rev} ne $couch_doc->rev) {
                Data::CouchDB::RevisionNotMatched->throw(
                    uri      => $self->log_uri,
                    db       => $self->db,
                    document => $doc,
                    error    => "CouchDB Revision mismatch " . $data->{_rev} . '!=' . $couch_doc->rev
                );
            }
            delete $data->{_rev};
        }

        delete $data->{_id} if (exists $data->{_id});

        try {
            $couch_doc->data($data);
            $couch_doc->update;
        }
        catch {
            Data::CouchDB::UpdateFailed->throw(
                uri      => $self->log_uri,
                db       => $self->db,
                document => $doc,
                error    => $_,
            );
        };

    }

    $data         = $couch_doc->data;
    $data->{_rev} = $couch_doc->rev;
    $data->{_id}  = $couch_doc->id;

    return $data;
}

=head2 view

Query a couchdb view

Usage,
    Without Parameters
        $couchdb->view($db, $viewname);

    With Parameters
        $couchdb->view($db, $viewname, $parameters);

        $parameters is a HashRef


=cut

sub view {
    my $self      = shift;
    my $view      = shift;
    my $params    = shift;

    return unless ($view);

    my $couch_design_doc = $self->_retrieved_design_doc;
    my $result;
    try { $result = $couch_design_doc->queryView($view, %$params); }
    catch {
        Data::CouchDB::QueryFailed->throw(
            db    => $self->db,
            uri   => $self->log_uri,
            view  => $view,
            error => $_,
        );
    };

    my $documents = [];
    foreach my $row (@{$result->{rows}}) {
        if ($row->{id}) {
            push @{$documents}, $row->{id};
        } else {
            push @{$documents}, $row;
        }
    }

    return $documents;
}

=head2 can_connect

Confirms that you can connect to this couchdb

Usage,
    if($couchdb->can_connect) {
        ...
    }

Returns,
    1     - can read
    undef - otherwise
=cut

sub can_connect {
    my $self = shift;

    my $client = CouchDB::Client->new(uri => $self->uri);
    return $client->testConnection;
}

=head2 create_or_update_view

Creates a CouchDB view.

=cut

sub create_or_update_view {
    my $self     = shift;
    my $new_view = shift;

    my $doc = $self->_design_doc;
    my $design_doc;
    try { $doc->retrieve; $design_doc = $doc->data; };

    # possibly new view
    unless ($design_doc) {
        $doc->create;
        $design_doc = {};
        $design_doc->{views} = {};
    }

    $design_doc->{views} = {%{$design_doc->{views}}, %$new_view};
    try {
        $doc->data($design_doc);
        $doc->update;
    }
    catch {
        Data::CouchDB::UpdateFailed->throw(
            uri      => $self->log_uri,
            db       => $self->db,
            document => $self->design_doc,
            error    => $_,
        );
    };

    # Recreate design doc since we updated it.
    $self->_design_doc($self->_build__design_doc);
    $self->_retrieved_design_doc($self->_build__retrieved_design_doc);

    return 1;
}

=head2 create_document

Creates a couch document

Usage,
    With known doc_id
        my $doc_id = $couchdb->create_document($db, $doc_id);

    Without known doc_id or with rev
        my $doc_id = $couchdb->create_document($db);

=cut

sub create_document {
    my $self      = shift;
    my $doc       = shift;

    my $couch_doc = $self->_doc($doc);

    try { $couch_doc->create; }
    catch {
        Data::CouchDB::Exception->throw(
            uri     => $self->log_uri,
            error   => $_,
            message => "Creating " . (($doc) ? $doc : "new") . " in " . $self->db . " failed: $_",
        );
    };

    return $couch_doc->id;
}

=head2 delete_document

Deletes a couch document

Usage,
    $couchdb->delete_document($db, $doc_id);


=cut

sub delete_document {
    my $self      = shift;
    my $doc       = shift;
    my $couch_doc = $self->_doc($doc);

    try {
        $couch_doc->retrieve;
        $couch_doc->delete;
    }
    catch {
        Data::CouchDB::Exception->throw(
            uri     => $self->log_uri,
            error   => $_,
            message => "Deleting $doc from " . $self->db . " failed",
        );
    };

    return 1;
}

=head2 create_database

Creates a CouchDB Database.

Usage,
    $couchdb->create_database($db);


=cut

sub create_database {
    my $self = shift;
    return $self->_db->create;
}

=head2 database_exists

=cut

sub database_exists {
    my $self = shift;
    return $self->_couchdb_client->dbExists($self->db);
}

sub _doc {
    my $self = shift;
    my $doc  = shift;

    return $self->_db->newDoc($doc);
}

has _couchdb_client => (
    is         => 'ro',
    isa        => 'CouchDB::Client',
    lazy_build => 1,
);

has _db => (
    is         => 'ro',
    isa        => 'CouchDB::Client::DB',
    lazy_build => 1,
);

has '_design_doc' => (
    is         => 'rw',
    lazy_build => 1,
);

has '_retrieved_design_doc' => (
    is         => 'rw',
    lazy_build => 1,
);

sub _build__couchdb_client {
    my $self = shift;

    my $params = {};
    $params->{uri} = $self->uri;
    $params->{ua} = $self->ua if ($self->ua);

    my $client = CouchDB::Client->new($params);
    unless ($client->testConnection) {
        Data::CouchDB::ConnectionFailed->throw(
            uri   => $self->log_uri,
            error => 'Connection to  ' . $self->log_uri . ' failed',
        );
    }

    return $client;
}

sub _build__db {
    my $self = shift;
    return $self->_couchdb_client->newDB($self->db);
}

sub _build__design_doc {
    my $self = shift;

    return $self->_db->newDesignDoc($self->design_doc);
}

sub _build__retrieved_design_doc {
    my $self = shift;

    my $design_doc = $self->_design_doc;
    try { $design_doc->retrieve; }
    catch {
        Data::CouchDB::RetrieveFailed->throw(
            db       => $self->db,
            uri      => $self->log_uri,
            document => $self->design_doc,
            error    => $_,
        );
    };

    return $self->_design_doc;
}

sub _build_uri {
    my $self     = shift;
    my $protocol = $self->protocol;
    my $url      = $protocol;

    # If couchdb is not set then it has no password
    if ($self->port ne 5984 and $self->couchdb) {
        $url .= 'couchdb:' . $self->couchdb . '@';
    }

    $url .= $self->host . ':' . $self->port . '/';
    return $url;
}

sub _build_log_uri {
    my $self = shift;

    return $self->protocol . $self->host . ':' . $self->port . '/';
}

sub _build_protocol {
    my $self = shift;
    my $port = $self->port;
    my $protocol;

    if ($port eq 6984) {
        $protocol = 'https://';
    } elsif ($port eq 5984) {
        $protocol = 'http://';
    }

    return $protocol;
}

sub _build_ua {
    my $self = shift;

    my $ua = LWP::UserAgent->new();
    $ua->ssl_opts(
        verify_hostname => 0,
        SSL_verify_mode => SSL_VERIFY_NONE
    );

    return $ua;
}

__PACKAGE__->meta->make_immutable;

1;

